File: src/main/java/io/pivotal/kotlindemo/java/accounts/DefaultAccountRepository.java
package io.pivotal.kotlindemo.java.accounts;

import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVParser;
import org.apache.commons.csv.CSVRecord;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static java.lang.Long.parseLong;

public class DefaultAccountRepository implements AccountRepository {

    private Map<Long, Account> accountMap = new HashMap<>();

    @Override
    public Account create(CreateAccountRequest createAccountRequest) {
        Long id = (long) accountMap.values().size() + 1;
        Account account = new Account(
            id,
            createAccountRequest.getClientId(),
            createAccountRequest.getNumber(),
            createAccountRequest.getFullname(),
            createAccountRequest.getNickname(),
            createAccountRequest.getType());

        accountMap.put(id, account);

        return account;
    }

    @Override
    public List<Account> findAll() {
        return new ArrayList<>(accountMap.values());
    }

    @Override
    public Account findById(Long id) {
        Account account = accountMap.get(id);
        if (account == null) {
            throw new NoAccountFoundException();
        }
        return account;
    }

    @Override
    public List<Account> findByClientId(Long id) {
        return accountMap
            .values()
            .stream()
            .filter(account -> account.getClientId().equals(id))
            .collect(Collectors.toList());
    }

    public void initialSeed() {
        ClassLoader classLoader = getClass().getClassLoader();
        InputStream inputStream = classLoader.getResourceAsStream("accounts.csv");
        InputStreamReader input = new InputStreamReader(inputStream);
        try {
            CSVParser csvParser = CSVFormat.RFC4180.withFirstRecordAsHeader().parse(input);
            for (CSVRecord record : csvParser) {
                Long id = parseLong(record.get("Id"));
                Long clientId = parseLong(record.get("Client Id"));
                String number = record.get("Number");
                String fullname = record.get("Full Name");
                String nickname = record.get("Nickname");
                AccountType type = AccountType.valueOf(record.get("Type"));

                Account account = new Account(
                    id,
                    clientId,
                    number,
                    fullname,
                    nickname,
                    type
                );
                accountMap.put(account.getId(), account);
            }
        } catch (IOException e) {
        }
    }

}

File: src/main/kotlin/io/pivotal/kotlindemo/kotlin/accounts/DefaultAccountRepository.kt
package io.pivotal.kotlindemo.kotlin.accounts

import org.apache.commons.csv.CSVFormat
import java.io.IOException
import java.io.InputStreamReader

class DefaultAccountRepository : AccountRepository {
    private val accountsMap = HashMap<Long, Account>()

    override fun create(createAccountRequest: CreateAccountRequest): Account {
        val id = accountsMap.size.toLong()
        val account = Account(
            id = id,
            clientId = createAccountRequest.clientId,
            number = createAccountRequest.number,
            fullname = createAccountRequest.fullname,
            nickname = createAccountRequest.nickname,
            type = createAccountRequest.type
        )
        accountsMap.put(id, account)
        return account
    }

    override fun findAll(): List<Account> {
        return accountsMap.values.toList()
    }

    override fun findById(id: Long): Account {
        return accountsMap[id] ?: throw NoAccountFoundException()
    }

    override fun findByClientId(clientId: Long): List<Account> {
        return accountsMap.values.filter { it -> it.clientId == clientId }
    }

    fun initialSeed() {
        val classLoader = javaClass.classLoader
        val inputStream = classLoader.getResourceAsStream("accounts.csv")
        val input = InputStreamReader(inputStream!!)
        try {
            val csvParser = CSVFormat.RFC4180.withFirstRecordAsHeader().parse(input)
            for (record in csvParser) {
                val id = record.get("Id").toLong()

                accountsMap[id] = Account(
                    id = id,
                    clientId = record.get("Client Id").toLong(),
                    number = record.get("Number"),
                    fullname = record.get("Full Name"),
                    nickname = record.get("Nickname"),
                    type = AccountType.valueOf(record.get("Type"))
                )
            }
        } catch (e: IOException) {
        }
    }

}
